pilot/pkg/security/authn/v1beta1/policy_applier.go

```diff
@@ -59,10 +63,73 @@ func (a *v1beta1PolicyApplier) JwtFilter(isXDSMarshalingToAnyEnabled bool) *http
 	return out
 }
 
+// All explaining code link can be removed before merging.
+func convertToIstioAuthnFilterConfig(jwtRules []*v1beta1.JWT) *authn_filter.FilterConfig {
+	p := authn_alpha.Policy{
+		// Targets are not referenced in proxy repo.
+		// https://github.com/istio/istio/blob/d8eb9d3699ef8945aa4cc6a1001a5455683df3b5/pilot/pkg/security/authn/v1alpha1/policy_applier.go#L265
+
+		// Origin will be optional since we don't reject req.
+		// Add Peers since we need that to trigger identity extraction in Authn filter.
+		// https://github.com/istio/proxy/blob/24e971ff31c5cce22e9ab49f8478629f50664846/src/envoy/http/authn/peer_authenticator.cc#L41
+		Peers: []*authn_alpha.PeerAuthenticationMethod{
+			&authn_alpha.PeerAuthenticationMethod{
+				Params: &authn_alpha.PeerAuthenticationMethod_Mtls{
+					Mtls: &authn_alpha.MutualTls{},
+				},
+			},
+		},
+		// https://github.com/istio/proxy/search?q=targets&unscoped_q=targets
+		OriginIsOptional: true,
+		// Peers not really needed, always invoke the
+		// https://github.com/istio/proxy/blob/24e971ff31c5cce22e9ab49f8478629f50664846/src/envoy/http/authn/http_filter.cc#L84
+		PeerIsOptional: true,
+		// Always bind request.auth.principal from JWT origin. In v2 policy, authorization config specifies what principal to
+		// choose from instead, rather than in authn config.
+		PrincipalBinding: authn_alpha.PrincipalBinding_USE_ORIGIN,
+	}
+
+	// TODO:(incfly): does it matter if this is empty but not nil?
+	jwtLocation := map[string]string{}
+	for _, jwt := range jwtRules {
+		p.Origins = append(p.Origins, &authn_alpha.OriginAuthenticationMethod{
+			Jwt: &authn_alpha.Jwt{
+				// https://github.com/istio/proxy/blob/24e971ff31c5cce22e9ab49f8478629f50664846/src/envoy/http/authn/authenticator_base.cc#L126
+				// used for getting the filter data.
+				// All other fields are irrelevant.
+				Issuer: jwt.GetIssuer(),
+			},
+		})
+		// Map to the same location.
+		jwtLocation[jwt.Issuer] = jwt.Issuer
+	}
+
+	return &authn_filter.FilterConfig{
+		Policy:                    &p,
+		JwtOutputPayloadLocations: jwtLocation,
+		// Same as before.
+		SkipValidateTrustDomain: features.SkipValidateTrustDomain.Get(),
+	}
+}
+
+// AuthNFilter returns the Istio authn filter config for a given authn Beta policy:
+// istio.authentication.v1alpha1.Policy policy, we specially constructs the old filter config to
+// ensure Authn Filter won't reject the request, but still transform the attributes, e.g. request.auth.principal.
+// TODO: confirm the proxyType does not matter.
 func (a *v1beta1PolicyApplier) AuthNFilter(proxyType model.NodeType, isXDSMarshalingToAnyEnabled bool) *http_conn.HttpFilter {
-	// TODO(diemtvu) implement this.
-	log.Errorf("AuthNFilter(%v, %v) is not yet implemented", proxyType, isXDSMarshalingToAnyEnabled)
-	return nil
+	out := &http_conn.HttpFilter{
+		Name: authn_model.AuthnFilterName,
+	}
+	filterConfigProto := convertToIstioAuthnFilterConfig(a.processedJwtRules)
+	if filterConfigProto == nil {
```

Comment:
nit. `convertToIstioAuthnFilterConfig` never returns nil (should it?), so this is not necessary.

Reply:
i think that's fine. caller be more agnostic of the actual implementation is a good thing IMO.